datatypes/src/raster/operations/map_masked_elements.rs

```diff
@@ -0,0 +1,234 @@
+                    out_no_data_count += 1;
+                }
+            });
+        // dbg!(in_no_data_count, out_no_data_count);
+
+        MaskedGrid::new(new_data, validity_mask)
+            .expect("Creation of grid with dimension failed before")
+    }
+}
+
+impl<In, Out, F, G> MapMaskedElements<In, Out, F> for GridOrEmpty<G, In>
+where
+    In: 'static + Copy,
+    Out: 'static + Copy + Default,
+    G: GridSize + Clone + PartialEq,
+    F: Fn(Option<In>) -> Option<Out>,
+{
+    type Output = GridOrEmpty<G, Out>;
+
+    fn map_or_mask_elements(self, map_fn: F) -> Self::Output {
+        match self {
+            GridOrEmpty::Grid(grid) => GridOrEmpty::Grid(grid.map_or_mask_elements(map_fn)),
+            GridOrEmpty::Empty(empty) => GridOrEmpty::Empty(empty.convert_dtype()),
+        }
+    }
+}
+
+impl<In, Out, F> MapMaskedElements<In, Out, F> for RasterTile2D<In>
+where
+    In: 'static + Copy + PartialEq,
+    Out: 'static + Copy + Default,
+    F: Fn(Option<In>) -> Option<Out>,
+{
+    type Output = RasterTile2D<Out>;
+
+    fn map_or_mask_elements(self, map_fn: F) -> Self::Output {
+        RasterTile2D {
+            grid_array: self.grid_array.map_or_mask_elements(map_fn),
+            time: self.time,
+            tile_position: self.tile_position,
+            global_geo_transform: self.global_geo_transform,
+            properties: self.properties,
+        }
+    }
+}
+
+impl<In, Out, F, G> MapMaskedElementsParallel<In, Out, F> for Grid<G, In>
+where
+    In: 'static + Copy + PartialEq + Send + Sync,
+    Out: 'static + Copy + Send + Sync + Default,
+    G: GridSize + Clone + Send + Sync + PartialEq,
+    F: Fn(Option<In>) -> Option<Out> + Sync + Send,
+{
+    type Output = MaskedGrid<G, Out>;
+    fn map_or_mask_elements_parallel(self, map_fn: F) -> Self::Output {
+        MaskedGrid::from(self).map_or_mask_elements_parallel(map_fn) // TODO: implement on & so that it does not clone the data
+    }
+}
+
+impl<In, Out, F, G> MapMaskedElementsParallel<In, Out, F> for MaskedGrid<G, In>
+where
+    In: 'static + Copy + PartialEq + Send + Sync,
+    Out: 'static + Copy + Send + Sync + Default,
+    G: GridSize + Clone + Send + Sync + PartialEq,
+    F: Fn(Option<In>) -> Option<Out> + Sync + Send,
+{
+    type Output = MaskedGrid<G, Out>;
+    fn map_or_mask_elements_parallel(self, map_fn: F) -> Self::Output {
+        let MaskedGrid {
+            inner_grid: data,
+            validity_mask,
+        } = self;
+
+        let shape = data.shape.clone();
+
+        let (new_data, new_mask): (Vec<Out>, Vec<bool>) = data
+            .data
+            .into_par_iter()
+            .with_min_len(data.shape.axis_size_x())
+            .zip(
+                validity_mask
+                    .data
+                    .into_par_iter()
+                    .with_min_len(validity_mask.shape.axis_size_x()),
+            )
+            .map(|(i, m)| {
+                let in_value = if m { Some(i) } else { None };
```

Comment:
No it is not clear. This is the way arrow does it. And we should test that. We could also use a bitmask instead of a groid of booleans. However, i think optimization of this is a new task